src/NUnitFramework/framework/Internal/Commands/SetUpTearDownItem.cs

```diff
@@ -92,11 +92,35 @@ private void RunSetUpOrTearDownMethod(TestExecutionContext context, IMethodInfo
             _methodValidator?.Validate(method.MethodInfo);
 
             var methodInfo = MethodInfoCache.Get(method);
+            RaiseStartedEvent(context, methodInfo);
 
-            if (methodInfo.IsAsyncOperation)
-                AsyncToSyncAdapter.Await(() => InvokeMethod(method, context));
-            else
-                InvokeMethod(method, context);
+            try
+            {
+                if (methodInfo.IsAsyncOperation)
+                    AsyncToSyncAdapter.Await(() => InvokeMethod(method, context));
+                else
+                    InvokeMethod(method, context);
+            }
+            finally
+            {
+                RaiseFinishedEvent(context, methodInfo);
+            }
+        }
+
+        private static void RaiseStartedEvent(TestExecutionContext context, MethodInfoCache.TestMethodMetadata methodInfo)
+        {
+            if (methodInfo.HasOneTimeSetUpAttribute)
```

Comment:
Why do you check this?  You should not be here if you were not in a Setup/Teardown method.

I assume you try to separate between pure OneTimeSetup/TD and the test Setup/TD.  That is checked earlier.  It might be necessary to state that information inside the SetUpTearDownItem.

Reply:
Right. Updated in latest commit.
Changed to distinguish OneTimeSetUp and SetUp with the TestContext.